model/src/payouts/quanto.rs

```diff
@@ -0,0 +1,609 @@
+        n_contracts: u64,
+        leverage_long: Leverage,
+        leverage_short: Leverage,
+        n_payouts: usize,
+        multiplier: Decimal,
+        fee_offset: CompleteFee,
+    ) -> Self {
+        Self {
+            n_payouts,
+            initial_price,
+            n_contracts,
+            leverage_long,
+            leverage_short,
+            multiplier,
+            fee_offset,
+        }
+    }
+
+    /// Discretize the payout curve into distinct payouts.
+    pub fn discretized_payouts(&self) -> Result<Vec<Payout>, Error> {
+        let initial_margin_long = self.initial_margin_long()?;
+        let initial_margin_short = self.initial_margin_short()?;
+        let initial_margin_total = initial_margin_long + initial_margin_short;
+
+        match self.fee_offset {
+            CompleteFee::LongPaysShort(long_owes) if long_owes > initial_margin_long => {
+                return Err(Error::LongOwesTooMuch {
+                    owing: long_owes,
+                    margin: initial_margin_long,
+                });
+            }
+            CompleteFee::ShortPaysLong(short_owes) if short_owes > initial_margin_short => {
+                return Err(Error::ShortOwesTooMuch {
+                    owing: short_owes,
+                    margin: initial_margin_short,
+                });
+            }
+            _ => {}
+        }
+
+        let long_liquidation_interval = self
+            .long_liquidation_interval()
+            .context("Could not calculate long liquidation interval")?;
+        let short_liquidation_interval = self
+            .short_liquidation_interval()
+            .context("Could not calculate short liquidation interval")?;
+
+        let long_liquidation_threshold = long_liquidation_interval.end();
+        let short_liquidation_threshold = short_liquidation_interval.start();
+
+        let long_liquidation_payout = Payout {
+            interval: long_liquidation_interval.clone(),
+            long: Amount::ZERO,
+            short: initial_margin_total,
+        };
+
+        let mut payouts = vec![long_liquidation_payout];
+
+        // We want to split the settlement region of the curve into `n-payouts - 2` segments
+        let step = Decimal::from(short_liquidation_threshold - long_liquidation_threshold)
+            / Decimal::from(self.n_payouts - 2);
+        let step = step.to_u64().context("Could not convert step to u64")?;
+
+        // We start building the settlement intervals right after the end of the long liquidation
+        // interval
+        let mut prev_upper_bound = *long_liquidation_threshold;
+        loop {
+            let lower_bound = prev_upper_bound + 1;
+            let upper_bound = lower_bound + step;
+
+            // The settlement region of the curve ends as soon as the sub-interval we are building
+            // either touches or overlaps with the start of the short liquidation interval
+            if upper_bound >= short_liquidation_threshold - 1 {
```

Comment:
Can there be a gap between `upper_bound` and `short_liquidation_threshold`? 

Thinking out loud here: 

If `lower_bound=0`, `step=10`, then ` upper_bound=10` and `short_liquidation_threshold=15`, then we only generate an interval from `0..10`.

Reply:
Ah, got it, you are treating it as a special case below :)